File: src/main/java/de/aquafun3d/bingo/utils/teams/BingoTeams.java
package de.aquafun3d.bingo.utils.teams;

import de.aquafun3d.bingo.utils.helpers.IHelpers;
import de.aquafun3d.bingo.utils.scoreboards.IScoreboards;
import de.aquafun3d.bingo.utils.scoreboards.Scoreboards;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.Team;

import java.util.HashMap;
import java.util.UUID;

public class BingoTeams implements ITeams{

    private final IScoreboards _boards;
    private final IHelpers _helper;
    private final HashMap<UUID, Team> _teams = new HashMap<>();
    private final HashMap<String, String> _teamnames = new HashMap<>();
    private int _teamsize = 1;
    //TODO add Scores

    public BingoTeams(Scoreboards boards, IHelpers helpers){
        _boards = boards;
        _helper = helpers;
        initTeamnames();
    }

    private void addPlayerToTeam(Player player, String teamname){
        for(Scoreboard board : _boards.getPlayerBoards().values()){
            board.getTeam(teamname).addEntry(player.getName());
            //TODO muss ich das bei allen Spielern auch machen
        }
        _teams.put(player.getUniqueId(), _boards.getPlayerBoards().get(player.getUniqueId()).getTeam(teamname));
    }

    public void joinTeam(Player player, String teamname) {
        UUID uuid = player.getUniqueId();
        Team team = _boards.getPlayerBoards().get(uuid).getTeam(teamname);
        if(teamname.equals("spec")){
            addPlayerToTeam(player,teamname);
        }
        if(team.getSize() >= _teamsize){
            _helper.send(player, ChatColor.RED + "Team is already full");
        }else{
            Team playerTeam = _boards.getPlayerBoards().get(uuid).getEntryTeam(player.getName());
            if(playerTeam != null && !playerTeam.getName().equals(teamname)){
                _teams.remove(uuid);
                _boards.getPlayerBoards().get(uuid).getTeam(teamname).removeEntry(playerTeam.getName());
            }
            addPlayerToTeam(player,teamname);
        }
    }

    private void initTeamnames(){
        _teamnames.put("spec", ChatColor.GRAY + "[Spec] ");
        _teamnames.put("white", ChatColor.WHITE + "[#1] ");
        _teamnames.put("orange", ChatColor.GOLD + "[#2] ");
        _teamnames.put("magenta", ChatColor.LIGHT_PURPLE + "[#3] ");
        _teamnames.put("lightblue", ChatColor.AQUA + "[#4] ");
        _teamnames.put("yellow", ChatColor.YELLOW + "[#5] ");
        _teamnames.put("lime", ChatColor.GREEN + "[#6] ");
        _teamnames.put("pink", ChatColor.RED + "[#7] ");
        _teamnames.put("gray", ChatColor.DARK_GRAY + "[#8] ");
        _teamnames.put("lightgray", ChatColor.GRAY + "[#9] ");
        _teamnames.put("cyan", ChatColor.DARK_AQUA + "[#10] ");
        _teamnames.put("purple", ChatColor.DARK_PURPLE + "[#11] ");
        _teamnames.put("blue", ChatColor.BLUE + "[#12] ");
        _teamnames.put("brown", ChatColor.GOLD + "[#13] ");
        _teamnames.put("green", ChatColor.DARK_GREEN + "[#14] ");
        _teamnames.put("red", ChatColor.DARK_RED + "[#15] ");
        _teamnames.put("black", ChatColor.DARK_GRAY + "[#16] ");
    }

    public String getTeamPrefix(String name){
        return _teamnames.get(name);
    }

    public Team getPlayerTeam(Player player) {
        return _teams.get(player.getUniqueId());
    }

    public String getPlayerTeamName(Player player) {
        return getPlayerTeam(player).getName();
    }

    public String getPlayerTeamPrefix(Player player) {
        return _teams.get(player.getUniqueId()).prefix().toString();
    }


    public final HashMap<UUID,Team> getTeams(){
        return this._teams;
    }
}

File: src/main/java/de/aquafun3d/bingo/utils/teams/BingoTeams.kt
package de.aquafun3d.bingo.utils.teams

import de.aquafun3d.bingo.utils.helpers.IHelpers
import de.aquafun3d.bingo.utils.helpers.ISettings
import de.aquafun3d.bingo.utils.scoreboards.IScoreboards
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.entity.Player
import org.bukkit.scoreboard.Team
import java.util.*

class BingoTeams(private val _boards: IScoreboards, private val _helper: IHelpers, private val _settings: ISettings) : ITeams {
    private val _teams = mutableMapOf<UUID, Team>()
    private val _teamnames = mutableMapOf<String, Component>()

    init {
        initTeamnames()
    }

    private fun addPlayerToTeam(player: Player, teamname: String) {
        for (board in _boards.getPlayerboards().values) {
            board.getTeam(teamname)!!.addEntry(player.name)
        }
        _teams[player.uniqueId] = _boards.getPlayerboards()[player.uniqueId]!!.getTeam(teamname)!!
    }

    override fun joinTeam(player: Player, teamname: String) {
        val uuid = player.uniqueId
        val team = _boards.getPlayerboards()[uuid]!!.getTeam(teamname)
        if (teamname == "spec") {
            addPlayerToTeam(player, teamname)
            return
        }
        if (team!!.size >= _settings.getTeamsize()) {
            _helper.send(player, Component.text("Team is already full", NamedTextColor.RED))
        } else {
            val playerTeam = _boards.getPlayerboards()[uuid]!!.getEntryTeam(player.name)
            if (playerTeam != null && playerTeam.name != teamname) {
                _teams.remove(uuid)
                _boards.getPlayerboards()[uuid]!!.getTeam(teamname)!!.removeEntry(playerTeam.name)
            }
            addPlayerToTeam(player, teamname)
        }
    }

    private fun initTeamnames() {
        _teamnames["spec"] = Component.text("[Spec] ",NamedTextColor.GRAY)
        _teamnames["white"] = Component.text("[#1] ",NamedTextColor.WHITE)
        _teamnames["orange"] = Component.text("[#2] ", NamedTextColor.GOLD)
        _teamnames["magenta"] = Component.text("[#3] ", NamedTextColor.LIGHT_PURPLE)
        _teamnames["lightblue"] = Component.text("[#4] ", NamedTextColor.AQUA)
        _teamnames["yellow"] = Component.text("[#5] ", NamedTextColor.YELLOW)
        _teamnames["lime"] = Component.text("[#6] ",NamedTextColor.GREEN)
        _teamnames["pink"] = Component.text("[#7] ", NamedTextColor.RED)
        _teamnames["gray"] = Component.text("[#8] ", NamedTextColor.DARK_GRAY)
        _teamnames["lightgray"] = Component.text("[#9] ", NamedTextColor.GRAY)
        _teamnames["cyan"] = Component.text("[#10] ", NamedTextColor.DARK_AQUA)
        _teamnames["purple"] = Component.text("[#11] ", NamedTextColor.DARK_PURPLE)
        _teamnames["blue"] = Component.text("[#12] ", NamedTextColor.BLUE)
        _teamnames["brown"] = Component.text("[#13] ", NamedTextColor.GOLD)
        _teamnames["green"] = Component.text("[#14] ", NamedTextColor.DARK_GREEN)
        _teamnames["red"] = Component.text("[#15] ", NamedTextColor.DARK_RED)
        _teamnames["black"] = Component.text("[#16] ", NamedTextColor.DARK_GRAY)
    }

    override fun getTeamPrefix(name: String): Component {
        return _teamnames[name]!!
    }

    override fun getPlayerTeam(player: Player): Team? {
        return _teams[player.uniqueId]
    }

    override fun getPlayerTeamName(player: Player): String {
        return getPlayerTeam(player)!!.name
    }

    override fun getPlayerTeamPrefix(player: Player): Component {
        return _teams[player.uniqueId]!!.prefix()
    }

    override fun getTeams(): Map<UUID, Team>{
        return _teams
    }

    override fun updateTeamSuffix(player: Player, itemCount: Int){
        if(_teams[player.uniqueId]!!.name == "spec") return
        _teams[player.uniqueId]!!.suffix(Component.text(" (" + ((_settings.getQuantity() * 9) - itemCount) + "/" + _settings.getQuantity() * 9 + ")", NamedTextColor.YELLOW))
    }
}
